ang/crmCaseType.js

```diff
@@ -460,31 +479,55 @@
     $scope.addRole = function(roles, roleName) {
+      var matchingRole;
+      // does the label we've been given match up to any relationship type, based on id (which is label)
+      var matchingRoles = _.filter($scope.relationshipTypeOptions, {id: roleName});
+      if (matchingRoles.length) {
+        matchingRole = matchingRoles.shift();
+      }
+      // if found, is the corresponding machine name in the list of existing roles for the case type
       var names = _.pluck($scope.caseType.definition.caseRoles, 'name');
-      if (!_.contains(names, roleName)) {
-        var matchingRoles = _.filter($scope.relationshipTypeOptions, {id: roleName});
-        if (matchingRoles.length) {
-          var matchingRole = matchingRoles.shift();
-          roles.push({name: roleName, displaylabel: matchingRole.text});
-        } else {
-           CRM.loadForm(CRM.url('civicrm/admin/reltype', {action: 'add', reset: 1, label_a_b: roleName}))
-            .on('crmFormSuccess', function(e, data) {
-              var newType = _.values(data.relationshipType)[0];
-              roles.push({name: newType.label_b_a, displaylabel: newType.label_a_b});
-              // Assume that the case role should be A-B but add both directions as options.
-              $scope.relationshipTypeOptions.push({id: newType.label_a_b, text: newType.label_a_b});
-              if (newType.label_a_b != newType.label_b_a) {
-                $scope.relationshipTypeOptions.push({id: newType.label_b_a, text: newType.label_b_a});
-              }
-              $scope.$digest();
-            });
+      if (matchingRole) {
+        if (!_.contains(names, matchingRole.xmlNameOtherDirection)) {
+          roles.push({name: matchingRole.xmlNameOtherDirection, displaylabel: matchingRole.text});
         }
+      } else {
+         // not a known relationship type, so create on-the-fly
+         CRM.loadForm(CRM.url('civicrm/admin/reltype', {action: 'add', reset: 1, label_a_b: roleName}))
+          .on('crmFormSuccess', function(e, data) {
+            var newType = _.values(data.relationshipType)[0];
+            roles.push({name: newType.name_b_a, displaylabel: newType.label_a_b});
+            // Assume that the case role should be A-B but add both directions as options.
+            var newRelTypeOption = {
+              xmlNameOtherDirection: newType.name_b_a,
+              xmlName: newType.name_a_b,
+              id: newType.name_a_b,
+              text: newType.label_a_b
+            };
+            $scope.relationshipTypeOptions.push(newRelTypeOption);
+            $scope.relationshipTypeOptionsAll.push(newRelTypeOption);
+            // Add the other direction if different.
+            if (newType.label_a_b != newType.label_b_a) {
+              newRelTypeOption = {
+                xmlNameOtherDirection: newType.name_a_b,
+                xmlName: newType.name_b_a,
+                id: newType.name_b_a,
+                text: newType.label_b_a
+              };
```

Comment:
Same issues as above, so it should be:

```js
              newRelTypeOption = {
                xmlName: newType.name_a_b,
                xmlNameOtherDirection: newType.name_b_a,
                id: newType.label_a_b,
                text: newType.label_b_a
              };
```

Reply:
I also realize that we should probably push to `$scope.defaultRelationshipTypeOptions` since there's no way to say that your new role should be default.

Reply:
My testing notes say this was pre-existing so I ignored it. But yes that seems doable.

Reply:
>  The fact that we can't think of a good name for it may be a sign that it doesn't really correspond to anything.

This turns out to be true! In the new PR there's just xmlName, and then id and text stay because select2 is expecting variables with those names.

> I also realize that we should probably push to `$scope.defaultRelationshipTypeOptions` since there's no way to say that your new role should be default.

In the new PR this ends up happening almost automatically, because that object gets combined.